mule-shaders-module/src/main/java/org/mule/module/shaders/api/ReflectiveAccessShader.java

```diff
@@ -0,0 +1,632 @@
+    shadeRequest.getUberJar().getParentFile().mkdirs();
+
+    JarOutputStream out = null;
+    try {
+      out = new JarOutputStream(new BufferedOutputStream(new FileOutputStream(shadeRequest.getUberJar())));
+      goThroughAllJarEntriesForManifestTransformer(shadeRequest, resources, manifestTransformer, out);
+
+      // CHECKSTYLE_OFF: MagicNumber
+      Multimap<String, File> duplicates = HashMultimap.create(10000, 3);
+      // CHECKSTYLE_ON: MagicNumber
+
+      shadeJars(shadeRequest, resources, transformers, remapper, out, duplicates);
+
+      // CHECKSTYLE_OFF: MagicNumber
+      Multimap<Collection<File>, String> overlapping = HashMultimap.create(20, 15);
+      // CHECKSTYLE_ON: MagicNumber
+
+      for (String clazz : duplicates.keySet()) {
+        Collection<File> jarz = duplicates.get(clazz);
+        if (jarz.size() > 1) {
+          overlapping.put(jarz, clazz);
+        }
+      }
+
+      // Log a summary of duplicates
+      logSummaryOfDuplicates(overlapping);
+
+      if (overlapping.keySet().size() > 0) {
+        showOverlappingWarning();
+      }
+
+      for (ResourceTransformer transformer : transformers) {
+        if (transformer.hasTransformedResource()) {
+          transformer.modifyOutputStream(out);
+        }
+      }
+
+      out.close();
+      out = null;
+    } finally {
+      IOUtil.close(out);
+    }
+
+    for (Filter filter : shadeRequest.getFilters()) {
+      filter.finished();
+    }
+  }
+
+  /**
+   * {@link InputStream} that can peek ahead at zip header bytes.
+   */
+  private static class ZipHeaderPeekInputStream extends PushbackInputStream {
+
+    private static final byte[] ZIP_HEADER = new byte[] {0x50, 0x4b, 0x03, 0x04};
+
+    private static final int HEADER_LEN = 4;
+
+    protected ZipHeaderPeekInputStream(InputStream in) {
+      super(in, HEADER_LEN);
+    }
+
+    public boolean hasZipHeader() throws IOException {
+      final byte[] header = new byte[HEADER_LEN];
+      super.read(header, 0, HEADER_LEN);
+      super.unread(header);
+      return Arrays.equals(header, ZIP_HEADER);
+    }
+  }
+
+  /**
+   * Data holder for CRC and Size.
+   */
+  private static class CrcAndSize {
+
+    private final CRC32 crc = new CRC32();
+
+    private long size;
+
+    CrcAndSize(InputStream inputStream) throws IOException {
+      load(inputStream);
+    }
+
+    private void load(InputStream inputStream) throws IOException {
+      byte[] buffer = new byte[BUFFER_SIZE];
+      int bytesRead;
+      while ((bytesRead = inputStream.read(buffer)) != -1) {
+        this.crc.update(buffer, 0, bytesRead);
+        this.size += bytesRead;
+      }
+    }
+
+    public void setupStoredEntry(JarEntry entry) {
+      entry.setSize(this.size);
+      entry.setCompressedSize(this.size);
+      entry.setCrc(this.crc.getValue());
+      entry.setMethod(ZipEntry.STORED);
+    }
+  }
+
+  private void shadeJars(ShadeRequest shadeRequest, Set<String> resources, List<ResourceTransformer> transformers,
```

Comment:
refactor to reduce complexity.

Reply:
Since this file was copied from `org.apache.maven.plugins.shade.DefaultShader` and we have since confirmed with legal that we should open source the project under Apache v2.0, it makes more sense to keep this file as close to the original as possible, in order to make it easier to maintain for future versions of the plugin.